src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEFileSaveWatcher.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.VS.Automation;
+using VSLangProj;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.TempPE
+{
+    /// <summary>
+    /// This class handles being notified whenever a source file is saved by the hierarchy
+    /// </summary>
+    [Export(typeof(IFileActionHandler))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class TempPEFileSaveWatcher : IFileActionHandler
+    {
+        private readonly VSBuildManager _buildManager;
+        private readonly Lazy<ITempPEBuildManager> _tempPEBuildManager;
+
+        [ImportingConstructor]
+        public TempPEFileSaveWatcher(
+            [Import(typeof(BuildManager))]VSBuildManager buildManager,
+            Lazy<ITempPEBuildManager> tempPEBuildManager)
+        {
+            _buildManager = buildManager;
+            _tempPEBuildManager = tempPEBuildManager;
+        }
+
+        public async Task<bool> TryHandleFileSavedAsync(IProjectTree tree, string newFilePath, bool saveAs)
+        {
+            // We don't want to be the one to realise the Lazy<ITempPEManager> so ignore changes in projects that don't need TempPE
+            if (!_buildManager.ProjectNeedsTempPE) return false;
+
+            string path = tree.BrowseObjectProperties.ItemName;
+            // Save As will come through as a file rename so we don't need to double handle
+            if (!saveAs && path != null)
+            {
+                // The magic of MEF means that the instance we get from .Value is the same as the build managers instance, so it has all of the info it needs
+                await _tempPEBuildManager.Value.TryFireTempPEDirtyAsync(path);
+            }
+
+            // we're just listening in here, not taking action, so return false always
```

Comment:
Anyone running before that handles save can prevent us from seeing this notification.

Reply:
Yes, that is true, and it would also prevent single file generators from running. I couldn't see any other way to find out about files changing. There is a data flow update when it happens, but there is no information about what file changed, so I'm using this to avoid having to keep a list of inputs in memory with their last write times, and then checking last write times on every update.

Even if there is an action handler that blocks this from running, when anything actually asks for the TempPE it will still be recompiled as necessary due to the file date checking on the input file and DLL.

Reply:
Can we chat about this? In the world of branch switches I don't think listening to Saves is the right approach.